## Chapter 4 - Exercise 2: Hãy thực hiện những yêu cầu liên quan tới Data Frame, và đối chiếu với kết quả cho trước

In [172]:
import numpy as np
import pandas as pd

In [173]:
# Câu 1: Cho dictionary như sau: 
dic_1 = {'X':[78,85,96,80,86], 'Y':[84,94,89,83,86], 'Z':[86,97,96,72,83]}

# Tạo dataframe df1 từ dic_1
df1 = pd.DataFrame(dic_1)

# In nội dung của dataframe df1
df1

X   Y   Z
0  78  84  86
1  85  94  97
2  96  89  96
3  80  83  72
4  86  86  83

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>X</th>
      <th>Y</th>
      <th>Z</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>78</td>
      <td>84</td>
      <td>86</td>
    </tr>
    <tr>
      <th>1</th>
      <td>85</td>
      <td>94</td>
      <td>97</td>
    </tr>
    <tr>
      <th>2</th>
      <td>96</td>
      <td>89</td>
      <td>96</td>
    </tr>
    <tr>
      <th>3</th>
      <td>80</td>
      <td>83</td>
      <td>72</td>
    </tr>
    <tr>
      <th>4</th>
      <td>86</td>
      <td>86</td>
      <td>83</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [174]:
# Câu 2:
# Cho dictionary exam_data như sau:
exam_data  = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
        'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5,np.nan, 8, 19],
        'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
        'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}

# Cho list labels như sau:
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

In [175]:
# Câu 2a: Tạo dataframe df2 từ exam_data, với index của dataframe là lables
df2 = pd.DataFrame(exam_data, index = labels)

# In nội dung của dataframe df2
df2

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James    NaN         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura    NaN         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>score</th>
      <th>attempts</th>
      <th>qualify</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>a</th>
      <td>Anastasia</td>
      <td>12.5</td>
      <td>1</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>b</th>
      <td>Dima</td>
      <td>9.0</td>
      <td>3</td>
      <td>no</td>
    </tr>
    <tr>
      <th>c</th>
      <td>Katherine</td>
      <td>16.5</td>
      <td>2</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>d</th>
      <td>James</td>
      <td>NaN</td>
      <td>3</td>
      <td>no</td>
    </tr>
    <tr>
      <th>e</th>
      <td>Emily</td>
      <td>9.0</td>
      <td>2</td>
      <td>no</td>
    </tr>
    <tr>
      <th>f</th>
      <td>Michael</td>
      <td>20.0</td>
      <td>3</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>g</th>
      <td>Matthew</td>
      <td>14.5</td>
      <td>1</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>h</th>
      <td>Laura</td>
      <td>NaN</td>
      <td>1</td>
      <td>no</td>
    </tr>
    <tr>
      <th>i</th>
      <td>Kevin</td>
      <td>8.0</td>
      <td>2</td>
      <td>no</td>
    </tr>
    <tr>
      <th>j</th>
      <td>Jonas</td>
      <td>19.0</td>
      <td>1</td>
      <td>yes</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [176]:
# Câu 2b: Xem thông tin (info()) của dataframe df2
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10 entries, a to j
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   name      10 non-null     object 
 1   score     8 non-null      float64
 2   attempts  10 non-null     int64  
 3   qualify   10 non-null     object 
dtypes: float64(1), int64(1), object(2)
memory usage: 400.0+ bytes


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>&lt;class 'pandas.core.frame.DataFrame'&gt;
Index: 10 entries, a to j
Data columns (total 4 columns):
name        10 non-null object
score       8 non-null float64
attempts    10 non-null int64
qualify     10 non-null object
dtypes: float64(1), int64(1), object(2)
memory usage: 400.0+ bytes
</pre>
  
</details>

In [177]:
# Câu 3: Tạo dataframe df3 từ df2, chỉ chứa 2 cột là name và score
df3 = df2[['name', 'score']].copy()

# Xem kiểu dữ liệu (type) và kích thước (shape) của df3
print(type(df3), '\n', df3.shape)

# Hiển thị các dòng dữ liệu đầu tiên (head) của df3
df3.head()

<class 'pandas.core.frame.DataFrame'> 
 (10, 2)


name  score
a  Anastasia   12.5
b       Dima    9.0
c  Katherine   16.5
d      James    NaN
e      Emily    9.0

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>&lt;class 'pandas.core.frame.DataFrame'&gt;
(10, 2)
</pre>
  
  <div class="output_subarea output_html rendered_html output_result"><div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>score</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>a</th>
      <td>Anastasia</td>
      <td>12.5</td>
    </tr>
    <tr>
      <th>b</th>
      <td>Dima</td>
      <td>9.0</td>
    </tr>
    <tr>
      <th>c</th>
      <td>Katherine</td>
      <td>16.5</td>
    </tr>
    <tr>
      <th>d</th>
      <td>James</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>e</th>
      <td>Emily</td>
      <td>9.0</td>
    </tr>
  </tbody>
</table>
</div></div>
  
</details>

In [178]:
# Câu 4: Hiển thị thông tin thống kê chung (describe) của dataframe df2
df2.describe(include = 'all')

name      score   attempts qualify
count          10   8.000000  10.000000      10
unique         10        NaN        NaN       2
top     Anastasia        NaN        NaN     yes
freq            1        NaN        NaN       5
mean          NaN  13.562500   1.900000     NaN
std           NaN   4.693746   0.875595     NaN
min           NaN   8.000000   1.000000     NaN
25%           NaN   9.000000   1.000000     NaN
50%           NaN  13.500000   2.000000     NaN
75%           NaN  17.125000   2.750000     NaN
max           NaN  20.000000   3.000000     NaN

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>score</th>
      <th>attempts</th>
      <th>qualify</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>count</th>
      <td>10</td>
      <td>8.000000</td>
      <td>10.000000</td>
      <td>10</td>
    </tr>
    <tr>
      <th>unique</th>
      <td>10</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>2</td>
    </tr>
    <tr>
      <th>top</th>
      <td>Matthew</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>freq</th>
      <td>1</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>5</td>
    </tr>
    <tr>
      <th>mean</th>
      <td>NaN</td>
      <td>13.562500</td>
      <td>1.900000</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>std</th>
      <td>NaN</td>
      <td>4.693746</td>
      <td>0.875595</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>min</th>
      <td>NaN</td>
      <td>8.000000</td>
      <td>1.000000</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>25%</th>
      <td>NaN</td>
      <td>9.000000</td>
      <td>1.000000</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>50%</th>
      <td>NaN</td>
      <td>13.500000</td>
      <td>2.000000</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>75%</th>
      <td>NaN</td>
      <td>17.125000</td>
      <td>2.750000</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>max</th>
      <td>NaN</td>
      <td>20.000000</td>
      <td>3.000000</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [179]:
# Câu 5: Tạo dataframe df4 từ df2, chỉ chứa 2 cột là name và score, và chỉ có các dòng 1, 3, 5, 6
df4 = df2[['name', 'score']].iloc[[1,3,5,6]]

#tương đương
df4_1 = df2.iloc[[1,3,5,6], [0,1]]  # 0 là column name, 1 là column score


# In nội dung của dataframe df4
print(df4)
df4_1

      name  score
b     Dima    9.0
d    James    NaN
f  Michael   20.0
g  Matthew   14.5


name  score
b     Dima    9.0
d    James    NaN
f  Michael   20.0
g  Matthew   14.5

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>score</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>b</th>
      <td>Dima</td>
      <td>9.0</td>
    </tr>
    <tr>
      <th>d</th>
      <td>James</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>f</th>
      <td>Michael</td>
      <td>20.0</td>
    </tr>
    <tr>
      <th>g</th>
      <td>Matthew</td>
      <td>14.5</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [180]:
# Câu 6: Từ dataframe df2, in các dòng có dữ liệu score bị null
print(df2[pd.isnull(df2).any(axis=1)]) #tất cả các row có chứa giá trị null

df2[df2['score'].isnull()]

    name  score  attempts qualify
d  James    NaN         3      no
h  Laura    NaN         1      no


name  score  attempts qualify
d  James    NaN         3      no
h  Laura    NaN         1      no

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>score</th>
      <th>attempts</th>
      <th>qualify</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>d</th>
      <td>James</td>
      <td>NaN</td>
      <td>3</td>
      <td>no</td>
    </tr>
    <tr>
      <th>h</th>
      <td>Laura</td>
      <td>NaN</td>
      <td>1</td>
      <td>no</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [181]:
# Câu 7: Từ dataframe df2, in các dòng có score > 15 và <=20
df2[(df2['score'] > 15) & (df2['score'] <= 20)]

name  score  attempts qualify
c  Katherine   16.5         2     yes
f    Michael   20.0         3     yes
j      Jonas   19.0         1     yes

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>score</th>
      <th>attempts</th>
      <th>qualify</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>c</th>
      <td>Katherine</td>
      <td>16.5</td>
      <td>2</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>f</th>
      <td>Michael</td>
      <td>20.0</td>
      <td>3</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>j</th>
      <td>Jonas</td>
      <td>19.0</td>
      <td>1</td>
      <td>yes</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [182]:
# Câu 8: Cập nhật điểm (score) ở dòng 'd' thành 18
df2.loc['d', 'score'] = 18
df2


name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James   18.0         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura    NaN         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>score</th>
      <th>attempts</th>
      <th>qualify</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>a</th>
      <td>Anastasia</td>
      <td>12.5</td>
      <td>1</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>b</th>
      <td>Dima</td>
      <td>9.0</td>
      <td>3</td>
      <td>no</td>
    </tr>
    <tr>
      <th>c</th>
      <td>Katherine</td>
      <td>16.5</td>
      <td>2</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>d</th>
      <td>James</td>
      <td>18.0</td>
      <td>3</td>
      <td>no</td>
    </tr>
    <tr>
      <th>e</th>
      <td>Emily</td>
      <td>9.0</td>
      <td>2</td>
      <td>no</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [183]:
# Câu 9: Cho biết điểm (score) nào có tần suất xuất hiện nhiều nhất trong df2, và in ra những dòng có điểm là tần suất xuất hiện nhiều nhất 
df2['score'].mode()
print(df2['score'].mode()[0])

# Cho biết có bao nhiêu dòng thỏa điều kiện?
df2[df2['score'] == df2['score'].mode()[0]]

9.0


name  score  attempts qualify
b   Dima    9.0         3      no
e  Emily    9.0         2      no

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>9.0
</pre>
    
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>score</th>
      <th>attempts</th>
      <th>qualify</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>b</th>
      <td>Dima</td>
      <td>9.0</td>
      <td>3</td>
      <td>no</td>
    </tr>
    <tr>
      <th>e</th>
      <td>Emily</td>
      <td>9.0</td>
      <td>2</td>
      <td>no</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [184]:
# Câu 10: Thêm dòng k có nội dung như sau: ['Suresh', 15.5, 1, 'yes'] vào df2
df2.loc['k'] = ['Suresh', 15.5, 1, 'yes']

# Hiển thị 5 dòng cuối cùng (tail) của df2.
df2.tail

<bound method NDFrame.tail of         name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James   18.0         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura    NaN         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes
k     Suresh   15.5         1     yes>

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>score</th>
      <th>attempts</th>
      <th>qualify</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>g</th>
      <td>Matthew</td>
      <td>14.5</td>
      <td>1</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>h</th>
      <td>Laura</td>
      <td>NaN</td>
      <td>1</td>
      <td>no</td>
    </tr>
    <tr>
      <th>i</th>
      <td>Kevin</td>
      <td>8.0</td>
      <td>2</td>
      <td>no</td>
    </tr>
    <tr>
      <th>j</th>
      <td>Jonas</td>
      <td>19.0</td>
      <td>1</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>k</th>
      <td>Suresh</td>
      <td>15.5</td>
      <td>1</td>
      <td>yes</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [185]:
# Câu 11a: Thêm dòng l có nội dung như sau: ['Janny', 12.5, 2, 'yes'] vào df2. 
df2.loc['l'] = ['Janny', 12.5, 2, 'yes']

# Hiển thị 5 dòng cuối cùng (tail) của df2.
df2.tail

<bound method NDFrame.tail of         name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James   18.0         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura    NaN         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes
k     Suresh   15.5         1     yes
l      Janny   12.5         2     yes>

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>score</th>
      <th>attempts</th>
      <th>qualify</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>h</th>
      <td>Laura</td>
      <td>NaN</td>
      <td>1</td>
      <td>no</td>
    </tr>
    <tr>
      <th>i</th>
      <td>Kevin</td>
      <td>8.0</td>
      <td>2</td>
      <td>no</td>
    </tr>
    <tr>
      <th>j</th>
      <td>Jonas</td>
      <td>19.0</td>
      <td>1</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>k</th>
      <td>Suresh</td>
      <td>15.5</td>
      <td>1</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>l</th>
      <td>Janny</td>
      <td>12.5</td>
      <td>2</td>
      <td>yes</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [186]:
# Câu 11b: Xóa bỏ dòng l của df2. 
#df2.drop('l', inplace = True) 

# Hiển thị lại 5 dòng cuối cùng (tail) của df2.
df2.tail()

name  score  attempts qualify
h   Laura    NaN         1      no
i   Kevin    8.0         2      no
j   Jonas   19.0         1     yes
k  Suresh   15.5         1     yes
l   Janny   12.5         2     yes

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>score</th>
      <th>attempts</th>
      <th>qualify</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>g</th>
      <td>Matthew</td>
      <td>14.5</td>
      <td>1</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>h</th>
      <td>Laura</td>
      <td>NaN</td>
      <td>1</td>
      <td>no</td>
    </tr>
    <tr>
      <th>i</th>
      <td>Kevin</td>
      <td>8.0</td>
      <td>2</td>
      <td>no</td>
    </tr>
    <tr>
      <th>j</th>
      <td>Jonas</td>
      <td>19.0</td>
      <td>1</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>k</th>
      <td>Suresh</td>
      <td>15.5</td>
      <td>1</td>
      <td>yes</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [187]:
# Câu 12: Sắp xếp df2 tăng dần theo điểm (score)
df2.sort_values('score')

name  score  attempts qualify
i      Kevin    8.0         2      no
b       Dima    9.0         3      no
e      Emily    9.0         2      no
a  Anastasia   12.5         1     yes
l      Janny   12.5         2     yes
g    Matthew   14.5         1     yes
k     Suresh   15.5         1     yes
c  Katherine   16.5         2     yes
d      James   18.0         3      no
j      Jonas   19.0         1     yes
f    Michael   20.0         3     yes
h      Laura    NaN         1      no

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>score</th>
      <th>attempts</th>
      <th>qualify</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>i</th>
      <td>Kevin</td>
      <td>8.0</td>
      <td>2</td>
      <td>no</td>
    </tr>
    <tr>
      <th>b</th>
      <td>Dima</td>
      <td>9.0</td>
      <td>3</td>
      <td>no</td>
    </tr>
    <tr>
      <th>e</th>
      <td>Emily</td>
      <td>9.0</td>
      <td>2</td>
      <td>no</td>
    </tr>
    <tr>
      <th>a</th>
      <td>Anastasia</td>
      <td>12.5</td>
      <td>1</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>g</th>
      <td>Matthew</td>
      <td>14.5</td>
      <td>1</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>k</th>
      <td>Suresh</td>
      <td>15.5</td>
      <td>1</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>c</th>
      <td>Katherine</td>
      <td>16.5</td>
      <td>2</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>d</th>
      <td>James</td>
      <td>18.0</td>
      <td>3</td>
      <td>no</td>
    </tr>
    <tr>
      <th>j</th>
      <td>Jonas</td>
      <td>19.0</td>
      <td>1</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>f</th>
      <td>Michael</td>
      <td>20.0</td>
      <td>3</td>
      <td>yes</td>
    </tr>
    <tr>
      <th>h</th>
      <td>Laura</td>
      <td>NaN</td>
      <td>1</td>
      <td>no</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [192]:
# Câu 13: Thêm cột result vào df2, 
# dựa vào dữ liệu của cột'score', nếu dòng nào có điểm >=10 thì giá trị của cột result = 1, ngược lại = 0
df2['result'] = df2['score'].map(lambda x: 1 if x >=10 else 0)

# In nội dung của dataframe df2
df2

name  score  attempts qualify  result
a  Anastasia   12.5         1     yes       1
b       Dima    9.0         3      no       0
c  Katherine   16.5         2     yes       1
d      James   18.0         3      no       1
e   Samantha    9.0         2      no       0
f    Michael   20.0         3     yes       1
g    Matthew   14.5         1     yes       1
h      Laura    NaN         1      no       0
i      Kevin    8.0         2      no       0
j      Jonas   19.0         1     yes       1
k     Suresh   15.5         1     yes       1
l      Janny   12.5         2     yes       1

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>score</th>
      <th>attempts</th>
      <th>qualify</th>
      <th>result</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>i</th>
      <td>Kevin</td>
      <td>8.0</td>
      <td>2</td>
      <td>no</td>
      <td>0</td>
    </tr>
    <tr>
      <th>b</th>
      <td>Dima</td>
      <td>9.0</td>
      <td>3</td>
      <td>no</td>
      <td>0</td>
    </tr>
    <tr>
      <th>e</th>
      <td>Emily</td>
      <td>9.0</td>
      <td>2</td>
      <td>no</td>
      <td>0</td>
    </tr>
    <tr>
      <th>a</th>
      <td>Anastasia</td>
      <td>12.5</td>
      <td>1</td>
      <td>yes</td>
      <td>1</td>
    </tr>
    <tr>
      <th>g</th>
      <td>Matthew</td>
      <td>14.5</td>
      <td>1</td>
      <td>yes</td>
      <td>1</td>
    </tr>
    <tr>
      <th>k</th>
      <td>Suresh</td>
      <td>15.5</td>
      <td>1</td>
      <td>yes</td>
      <td>1</td>
    </tr>
    <tr>
      <th>c</th>
      <td>Katherine</td>
      <td>16.5</td>
      <td>2</td>
      <td>yes</td>
      <td>1</td>
    </tr>
    <tr>
      <th>d</th>
      <td>James</td>
      <td>18.0</td>
      <td>3</td>
      <td>no</td>
      <td>1</td>
    </tr>
    <tr>
      <th>j</th>
      <td>Jonas</td>
      <td>19.0</td>
      <td>1</td>
      <td>yes</td>
      <td>1</td>
    </tr>
    <tr>
      <th>f</th>
      <td>Michael</td>
      <td>20.0</td>
      <td>3</td>
      <td>yes</td>
      <td>1</td>
    </tr>
    <tr>
      <th>h</th>
      <td>Laura</td>
      <td>NaN</td>
      <td>1</td>
      <td>no</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [190]:
# Câu 14: Trong df2, thay tên 'Emily' thành 'Samantha'
df2['name'] = df2['name'].replace('Emily', 'Samantha')

# In lại nội dung của dataframe df2
df2

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James   18.0         3      no
e   Samantha    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura    NaN         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes
k     Suresh   15.5         1     yes
l      Janny   12.5         2     yes

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>score</th>
      <th>attempts</th>
      <th>qualify</th>
      <th>result</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>i</th>
      <td>Kevin</td>
      <td>8.0</td>
      <td>2</td>
      <td>no</td>
      <td>0</td>
    </tr>
    <tr>
      <th>b</th>
      <td>Dima</td>
      <td>9.0</td>
      <td>3</td>
      <td>no</td>
      <td>0</td>
    </tr>
    <tr>
      <th>e</th>
      <td>Samantha</td>
      <td>9.0</td>
      <td>2</td>
      <td>no</td>
      <td>0</td>
    </tr>
    <tr>
      <th>a</th>
      <td>Anastasia</td>
      <td>12.5</td>
      <td>1</td>
      <td>yes</td>
      <td>1</td>
    </tr>
    <tr>
      <th>g</th>
      <td>Matthew</td>
      <td>14.5</td>
      <td>1</td>
      <td>yes</td>
      <td>1</td>
    </tr>
    <tr>
      <th>k</th>
      <td>Suresh</td>
      <td>15.5</td>
      <td>1</td>
      <td>yes</td>
      <td>1</td>
    </tr>
    <tr>
      <th>c</th>
      <td>Katherine</td>
      <td>16.5</td>
      <td>2</td>
      <td>yes</td>
      <td>1</td>
    </tr>
    <tr>
      <th>d</th>
      <td>James</td>
      <td>18.0</td>
      <td>3</td>
      <td>no</td>
      <td>1</td>
    </tr>
    <tr>
      <th>j</th>
      <td>Jonas</td>
      <td>19.0</td>
      <td>1</td>
      <td>yes</td>
      <td>1</td>
    </tr>
    <tr>
      <th>f</th>
      <td>Michael</td>
      <td>20.0</td>
      <td>3</td>
      <td>yes</td>
      <td>1</td>
    </tr>
    <tr>
      <th>h</th>
      <td>Laura</td>
      <td>NaN</td>
      <td>1</td>
      <td>no</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div>
  
</details>

In [207]:
# Câu 15: Duyệt df2, in name, score, result: nếu nếu giá trị cột result = 1 thì in 'Pass', ngược lại thì in 'Fail'
for index, row in df2.iterrows():
    print(row['name'], row['score'], 'Result:', 'Pass' if row.result == 1 else 'Fail')

Anastasia 12.5 Result: Pass
Dima 9.0 Result: Fail
Katherine 16.5 Result: Pass
James 18.0 Result: Pass
Samantha 9.0 Result: Fail
Michael 20.0 Result: Pass
Matthew 14.5 Result: Pass
Laura nan Result: Fail
Kevin 8.0 Result: Fail
Jonas 19.0 Result: Pass
Suresh 15.5 Result: Pass
Janny 12.5 Result: Pass


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Kevin  - Score: 8.0  - Result: Fail
Dima  - Score: 9.0  - Result: Fail
Samantha  - Score: 9.0  - Result: Fail
Anastasia  - Score: 12.5  - Result: Pass
Matthew  - Score: 14.5  - Result: Pass
Suresh  - Score: 15.5  - Result: Pass
Katherine  - Score: 16.5  - Result: Pass
James  - Score: 18.0  - Result: Pass
Jonas  - Score: 19.0  - Result: Pass
Michael  - Score: 20.0  - Result: Pass
Laura  - Score: nan  - Result: Fail
</pre>
  
</details>